Add the ability to validate Chart in root directory

VannTen opened the pull request:
This modify the way to find Charts directory to be able to validate a 'single chart' repo, where the helm chart is at the root.
- instead of stopping at the first error, it still process the rest of the charts/values.
- allow to skip directories based by regex

EDIT:
- use the json output from kubeconform to get more readable logs

VannTen committed:
feat: allow validating a single chart directory

shivjm commented on main.go:
```diff
@@ -84,63 +86,62 @@ func main() {
-			fileLogger.Info().Msgf("Output: %s", output)
+			logger.Warn().Err(err).Msg("skipping path (error)")
+
+		} else if dirent.IsDir() && regexp.MustCompile("^"+cfg.RegexSkipDir+"$").MatchString(dirent.Name()) {
```
Is it possible to move the regexp out of the inner function and use [`QuoteMeta`](https://pkg.go.dev/regexp#QuoteMeta)?

VannTen replied:
Moving the regex out, absolutely. Regarding QuoteMeta, though, isn't that going to prevent people to use this as a regex ? (I'm realising that `.git` should be `\.git` ...).
At first I was thinking to have a list of litteral string rather than a regex, quoting the regex would prevent doing things such as `.*` or `skip_this_one|and_this|and_this` right ?

Aankhen replied:
I completely forgot that it’s *meant* to be a regex! 🤦‍♂️ Sorry about that. Ignore me.

shivjm replied:
I completely forgot that it’s *meant* to be a regex! 🤦‍♂️ Ignore me.

shivjm commented on main.go:
```diff
@@ -84,63 +86,62 @@ func main() {
 
+	err := filepath.WalkDir(cfg.ChartsDirectory.path, func(path string, dirent fs.DirEntry, err error) error {
+		logger := log.With().Str("path", path).Logger()
 		if err != nil {
```
Could we turn this into a series of early returns to reduce the rightward drift? Something like:

```go
if err != nil {
  logger.Warn().Err(err).Msg("skipping path (error)")
  return nil
}

if dirent.isDir() && skipRegex.MatchString(dirent.Name()) {
  logger.Info().Msg(…)
  return nil
}

if dirent.Name() != "Chart.yaml" && dirent.Name() != "Chart.yml" {
  return nil
}

chartDir := filepath.Dir(…)
// …
return fs.SkipDir
```

shivjm commented on main.go:
```diff
@@ -84,63 +86,62 @@ func main() {
-	return foreachChart(cfg.ChartsDirectory.path, func(base string) error {
-		logger := log.With().Str("chart", filepath.Base(base)).Logger()
-		valuesFiles, err := os.ReadDir(filepath.Join(base, TestsPath))
+	var validations_errors []error
```
Could we use `camelCase` for the new variable names too, like `validationsErrors`?

shivjm reviewed (changes requested):
Thanks for the PR. Am I right in thinking there are in fact three new features? Namely:

1. Validating single-chart repositories.
2. Not stopping at the first validation failure.
3. Skipping directories.

These are all very useful additions. I’m not sure about the logic of the directory skipping, though. Let’s say my repository contains a directory `foo` and a subdirectory `bar` inside that one. If I pass `regexSkipDir: foo`, should it still process `bar`? It seems to me—and maybe this is just my unfamiliarity with Go speaking—that since this code doesn’t return `SkipDir` when the regex matches, `bar` *will* be processed.

As far as the structure goes, do you think you could split this into three commits (one per feature)?

VannTen commented:
> Thanks for the PR. Am I right in thinking there are in fact three new features? Namely:
> 
>     1. Validating single-chart repositories.
> 
>     2. Not stopping at the first validation failure.
> 
>     3. Skipping directories.
> 
> 

Absolutely. I was hacking on this somewhat randomly to make it work with my repository and it organically evolved into this ^. 

> These are all very useful additions. I’m not sure about the logic of the directory skipping, though. Let’s say my repository contains a directory `foo` and a subdirectory `bar` inside that one. If I pass `regexSkipDir: foo`, should it still process `bar`? It seems to me—and maybe this is just my unfamiliarity with Go speaking—that since this code doesn’t return `SkipDir` when the regex matches, `bar` _will_ be processed.

You're totally right, that should return `fs.SkipDir`, I'll fix that. I got confused by the behavior depending on whether path is a dir or not.


> As far as the structure goes, do you think you could split this into three commits (one per feature)?

Will do :+1:

VannTen committed:
feat: validate all charts and report all failures

VannTen committed:
feat: allow skipping directories

Add a new parameters to skip searching charts under some directories.
We default to `\.git` as this is the default useful thing when running
on the root of a git reposository.

VannTen committed:
fix: simplify logging

VannTen committed:
feat: propagate JSON output from kubeconform upon validation error

shivjm reviewed (approved):
Looks good to me. I’m just going to tweak the commit messages to use [the Conventional Commits style](https://www.conventionalcommits.org/en/v1.0.0/) so they match the rest of the repository; don’t be alarmed by the different descriptions.

Thank you again for the PR. These features will come in very handy.

shivjm commented:
[Released in v0.3.0.](https://github.com/shivjm/helm-kubeconform-action/releases/tag/v0.3.0)

VannTen commented:
No problem, and thanks !
